### DISCUSSION TITLE
ChainablePromiseElement Type error

### DISCUSSION THREAD

**Discussion by krististepanek** [webdriverio/webdriverio#13940] [Q&A] [answered]
Implemented Typescript and ESLint in my WebdriverIO project. I have added `@wdio/types` into my package.json and added that and `@wdio/globals/types` in my tsconfig.json. I am using the `$` locator to find elements, but when I try to add a typing to the elements, I just see this error from ESLint `'ChainablePromiseElement' is not defined.eslint[no-undef](https://eslint.org/docs/latest/rules/no-undef)`. But if I try to define it as a different type, it says `Type 'ChainablePromiseElement' is not assignable to type 'Element'.` or `Type 'ChainablePromiseElement' is missing the following properties from type 'Promise<Element>': then, catch, finally, `
I want to type the elements as ChainablePromiseElements, I just can't figure out how to get rid of that linting error!

pageObject.ts
```  
buttonGetStarted(): ChainablePromiseElement {
    return $('~button_get_started_landing');
  }
```

tsconfig.json:
```
{
  "compilerOptions": {
    "target": "ES6",
    "module": "commonjs",
    "esModuleInterop": true,
    "strict": true,
    "outDir": "./dist",
    "paths": {
      "testData/*": ["./test/testData/*"],
      "iosPages/*": ["./test/ios/page/*"],
      "androidPages/*": ["./test/android/page/*"]
    },
    "types": [
      "node", 
      "@wdio/globals/types", 
      "@wdio/mocha-framework", 
      "@wdio/appium-service",
      "@wdio/browserstack-service",
      "@wdio/types"
    ]
  },
  "include": [
    "./test/**/*.ts"
  ]
}
```

eslint.config.mjs
```
import globals from 'globals';
import pluginJs from '@eslint/js';
import stylistic from '@stylistic/eslint-plugin';
import mochaPlugin from 'eslint-plugin-mocha';
import * as wdio from 'eslint-plugin-wdio';
import typescriptParser from '@typescript-eslint/parser';

export default [
  pluginJs.configs.recommended,
  mochaPlugin.configs.flat.recommended,
  wdio.configs['flat/recommended'],
  stylistic.configs.customize({
    indent: 2,
    quotes: 'single',
    semi: true,
    braceStyle: '1tbs',
    commaDangle: 'never'
  }),
  {
    plugins: {
      '@stylistic': stylistic
    },
    files: ['**/*.ts'],
    languageOptions: {
      parser: typescriptParser
    },
    rules: {
      'no-unused-vars': ['warn']
    }
  },
  {
    languageOptions: {
      globals: {
        ...globals.browser,
        ...globals.node
      }
    }
  }
];
```

**Comment by krististepanek** [comment, marked as answer]
Oof I finally got it by somewhere finding the correct thing to import, which looks obvious when you read it but was really hard to discover the right import
`import { ChainablePromiseElement } from 'webdriverio';`

Now I would love to be able to import this in a single spot instead of importing it in every page object file, but that's a different question I guess.

**Reply by christian-bromann** [reply]
Users should not deal with `ChainablePromiseElement` types in v9 it should be all just `WebdriverIO.Element`.

**Reply by krististepanek** [reply]
Thanks for following up, how do I import that? I am using Appium and Browserstack for a native mobile app, if that is relevant. Sorry if this is an obvious question, I'm just trying to get everything working without typescript or linting errors and I can't find a good example.

**Reply by christian-bromann** [reply]
Again, you are not supposed to import `ChainablePromiseElement`, which code is failing for you atm?

**Reply by krististepanek** [reply]
If I don't import anything in my page object file, I just get this eslint error `'WebdriverIO' is not defined.`
```
class BasePage {
  buttonGetStarted(): WebdriverIO.Element {
    return $('~button_get_started_landing');
  }
}
```

if I add the line `import { WebdriverIO } from '@wdio/types/build/Options';` to the top of the file, then I get the typescript error 
```
Type 'ChainablePromiseElement' is not assignable to type 'Element'.
  Types of property 'parent' are incompatible.
    Type 'Promise<Element | Browser | MultiRemoteBrowser>' is not assignable to type 'Element | Browser'.ts(2322)
```

**Reply by christian-bromann** [reply]
Can you create a minimal reproduction case?

**Reply by krististepanek** [reply]
Here's a [minimal setup](https://github.com/krististepanek/wdio-ts-error), when I open Base.page.ts in VSCode, I get these errors: 

![Screenshot 2024-12-04 at 11 50 09 AM](https://github.com/user-attachments/assets/842aad3a-5906-4e1c-9b2c-f4194098234d)

I expect I must be missing something in my eslint.config or tsconfig but I just can't figure out what

**Reply by christian-bromann** [reply]
> I expect I must be missing something in my eslint.config

Yes, you should disable the `no-undef` rule for TypeScript files.

The other issue will be taken care of in #13945, contributions are welcome!
